tests/storage/conftest.py

```diff
@@ -1,7 +1,66 @@
+from __future__ import annotations
+
+import pytest
+
+import logging
+
+# explicit import for package-scope fixtures
+from pkgfixtures import pool_with_saved_yum_state
+
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from lib.host import Host
+
+QCOW2_REPO = "qcow2"
+QCOW2_REPO_FILE = f"/etc/yum.repos.d/xcp-ng-{QCOW2_REPO}.repo"
+
+
 def pytest_collection_modifyitems(config, items):
     # modify ordering so that ext is always tested first,
     # before more complex storage drivers
     for item in reversed(list(items)):
         if "_ext_" in item.path.name:
             items.remove(item)
             items.insert(0, item)
+
+
+@pytest.fixture(scope='package')
+def pool_with_qcow2(hostA1: Host, image_format):
+    import concurrent.futures
+    pool = hostA1.pool
+
+    # If we're testing VHD image format, skip installing qcow2 and just return the pool.
+    # Use yield+return because this is a generator fixture.
+    if image_format == "vhd":
+        yield pool
+        return
+
+    def check_qcow2_installed(host):
+        if host.file_exists(QCOW2_REPO_FILE):
+            raise Exception(
+                f'{QCOW2_REPO_FILE} is already installed on host {host}. This should not be the case.'
+            )
+
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        list(executor.map(check_qcow2_installed, pool.hosts))
+
+    def install_qcow2(host):
+        logging.info(f"Installing qcow2 packages on host {host}...")
+        host.add_xcpng_repo(QCOW2_REPO, base_repo="vates")
+        host.yum_install(["sm", "sm-fairlock", "blktap"])
+        host.restart_toolstack(verify=True)
+
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        list(executor.map(install_qcow2, pool.hosts))
+
+    yield pool
+
+    def remove_qcow2(host):
+        logging.info(f"Cleaning up qcow2 packages from host {host}...")
+        host.remove_xcpng_repo(QCOW2_REPO)
+        host.yum_downgrade(["sm", "sm-fairlock", "blktap"])
```

Comment:
This is going to leave the hosts in a state that is not necessarily the initial state. You don't know if the contents enabled repositories match what we decided to install on the hosts to run the tests on it.

In addition to this, all this code will soon have to be deleted, once qcow2 packages are moved to the main testing repositories. That might be acceptable in order to facilitate running qcow2

Reply:
So, restoring yum saved state would be better in such case.

Once the qcow2 packages are moved to the main testing repo, we’ll still need to install them during tests or manually pre-tests, right? Or is the plan to merge the packages so that no separate installation of qcow2-related packages is needed?

Reply:
As mentioned earlier, it is assumed that qcow2 is installed and no specific package management is expected from the test, so removing this as well.